src/execution_plan/optimizations/optimize_cartesian_product.c

```diff
@@ -7,115 +7,114 @@
-#include "optimizations_util.h"
+#include "../../util/rax_extensions.h"
+#include "../../util/qsort.h"
 
-// Collects all consecutive filters beneath given op.
-static OpFilter **_locate_filters(OpBase *cp) {
+#define RAX_LT(a ,b) ((raxSize((a)->entities)) < (raxSize((b)->entities)))
+
+// This struct is an auxilary struct for sorting filters according to their referenced entities count.
+typedef struct {
+	OpFilter *filter;   // Filter operation
+	rax *entities;      // Contains the entities that the filter references.
+} Filter_And_Entities;
+
+// Collects all consecutive filters beneath given op, sort them by the number of referenced entities.
+static Filter_And_Entities *_locate_filters_and_entities(OpBase *cp) {
 	OpBase *parent = cp->parent;
-	OpFilter **filters = array_new(OpFilter *, 0);
+	Filter_And_Entities *filters = array_new(Filter_And_Entities, 0);
 
 	while(parent && parent->type == OPType_FILTER) {
+		Filter_And_Entities fae;
 		OpFilter *filter_op = (OpFilter *)parent;
-		filters = array_append(filters, filter_op);
+		fae.filter = filter_op;
+		// Collect referenced entities.
+		fae.entities = FilterTree_CollectModified(filter_op->filterTree);
+		filters = array_append(filters, fae);
 		parent = parent->parent;
 	}
-
+	// Sort by the number of referenced entities.
+	QSORT(Filter_And_Entities, filters, array_len(filters), RAX_LT);
 	return filters;
 }
 
-/**
- * @brief  Creates a structure of dual branched catesian product, followed by a filter.
- *         The generated record from the cartesian product completely satisfies the filter entities.
- * @param  *plan: Execution plan to associate its mapping with the new CP operation.
- * @param  *lhs: Left side branch.
- * @param  *rhs: Right side branch.
- * @param  *filter: Trailing filter operation.
- * @retval Pointer to the filter operation.
- */
-static OpBase *_chain_cp_and_filter(const ExecutionPlan *plan, OpBase *lhs, OpBase *rhs,
-									OpBase *filter) {
-	OpBase *cp = NewCartesianProductOp(plan);
-	ExecutionPlan_AddOp(cp, lhs);
-	ExecutionPlan_AddOp(cp, rhs);
-	ExecutionPlan_AddOp(filter, cp);
-	return filter;
+// Finds all the cartesian product's children which solve a specific filter entities.
+static OpBase **_find_entities_solving_branches(rax *entities, OpBase *cp) {
+	int entities_count = raxSize(entities);
+	OpBase **solving_branches = array_new(OpBase *, 1);
+	for(int i = 0; i < cp->childCount; i++) {
+		OpBase *branch = cp->children[i];
+		/* Locate references reduces the amount of entities upon each call
+		 * that partially solves the references. */
+		ExecutionPlan_LocateReferences(branch, NULL, entities);
+		int new_entities_count = raxSize(entities);
+		if(new_entities_count != entities_count) {
+			// Update entity count.
+			entities_count = new_entities_count;
+			// Add partially solving branch into the array.
+			solving_branches = array_append(solving_branches, branch);
+			// No need to continue from here.
+			if(entities_count == 0) break;
```

Comment:
move this condition into the for loop
`for(int i = 0; i < cp->childCount && entities_count > 0; i++)`

Reply:
done